### PR TITLE
fix: add null safety for extender apiResponse/items edge cases

### LINKED ISSUE #4192: Edge case with USS list of directories with items: undefined/null
**Describe the bug**

Review of another PR pointed out an edge case with USS list of directories with items: undefined/null and missing test coverage for edge case

- Issue: The new USS directory-detection heuristic is better than main for child-only extender listings, but still requires at least one returned item to classify as directory.  
- a residual edge case for successful empty-directory responses (`items: []`) from extenders.  
- Why it matters: Empty folders from some extenders may still fail to be cached/recognized consistently.  
- The unresolved edge case can regress silently in future refactors.  
- Suggested fix: Add an explicit empty-list fallback path for directory context (or a follow-up stat/lookup check) so `items: []` can still be treated correctly when appropriate. Add one unit test for successful empty-list directory response from an extender compatibility scenario.

**To Reproduce**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

- OS:
- Zowe Explorer Version:
- (Optional) Zowe CLI Version:
- (Optional) Do you have secure credentials enabled?
- (Optional) Are you using a custom credential manager?

**Additional context**

### PR DESCRIPTION
**Pull request by AarishMansur**
## Proposed changes

This PR fixes null safety gaps in USS file handling when extenders return responses without `apiResponse` or with `undefined`/`null` `items`. Currently, spreading `response.apiResponse` or calling `.filter()` / `.length` on undefined `items` throws a TypeError, which gets caught downstream and surfaces to the user as a generic error instead of showing an empty directory.

The fixes add `?? {}` and `?? []` fallbacks plus optional chaining (`?.`) in three files, and add unit tests for the empty directory extender scenario.

Fixes:  #4192 

## Release Notes

Milestone:  N/A

Changelog: Fixed crashes in USS file listing when extenders return undefined or missing apiResponse/items

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change which adds or improves functionality)
- [ ] Breaking change (a change that would cause existing functionality to not work as expected)
- [ ] Documentation (Markdown, README updates)
- [ ] Other (please specify above in "Proposed changes" section)

## Checklist

**All checks must pass before this pull request is reviewed by the Zowe Explorer squad.**

**General**

- [x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/zowe-explorer-vscode/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] All PR dependencies have been merged and published (if applicable)
- [ ] A GIF or screenshot is included in the PR for visual changes
- [x] The pre-publish command has been executed:
  - **v2 and below:** `yarn workspace vscode-extension-for-zowe vscode:prepublish`
  - **v3:** `pnpm --filter vscode-extension-for-zowe vscode:prepublish`
- [ ] New ZE APIs are tested with extender types that haven't adopted yet to determine breaking changes. Can use Zowe zFTP marketplace extension.

**Code coverage**

- [x] There is coverage for the code that I have added
- [x] I have added new unit test cases and they are passing
- [ ] I have added at least one end-to-end test for new features to validate behavior (if applicable)
- [x] I have manually tested the changes

**Deployment**

- [ ] I have tested new functionality with the FTP extension and profile verifying no extender profile type breakages introduced
- [ ] I have added developer documentation (if applicable)
- [ ] I have verified that all dependency updates (such as Zowe SDKs) in this pull request are compatible
- [ ] Documentation should be added to Zowe Docs
  - If you're an outside contributor, please post in the [#zowe-doc Slack channel](https://openmainframeproject.slack.com/archives/CC961JYMQ) to coordinate documentation.
  - Otherwise, please check with the rest of the squad about any needed documentation before merging.
- [ ] These changes may need ported to the appropriate branches (list here):

## Further comments

no this is not a complex solution  I used `?.` and `??` to guard the spots where `apiResponse` or `items` could be undefined

### CODE REVIEW

File: packages/zowe-explorer/src/trees/uss/UssFSProvider.ts [hunk 1]
```diff
@@ -309,7 +309,7 @@ export class UssFSProvider extends BaseProvider implements vscode.FileSystemProv
         return {
             ...response,
             apiResponse: {
-                ...response.apiResponse,
+                ...(response.apiResponse ?? {}),
                 items: (response.apiResponse.items ?? []).filter(keepRelative ? Boolean : (it): boolean => !/^\.{1,3}$/.test(it.name as string)),
             },
```
Comment (zFernand0) [inline packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:314]: should we just wrap this whole thing in a ternary?
for example:
```
            apiResponse: response.apiResponse ? {
                ...response.apiResponse,
                items: (response.apiResponse.items ?? []).filter(keepRelative ? Boolean : (it): boolean => !/^\.{1,3}$/.test(it.name as string)),
            } : {},
```

Comment (AarishMansur) [reply packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:314]: Thanks for review  and yes we can wrap it  with ternary but ternary would drop the `items` key entirely so thats why I used  `...(response.apiResponse ?? {})` to keep `apiResponse.items` as an array even when `apiResponse` is missing but if you prefer ternary I am happy to switch

Comment (zFernand0) [reply packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:314]: oh, I see.
then we might want ot add a `?.` in here:
- `items: (response.apiResponse.items ?? []).`
because if `response.apiResponse` is indeed undefined, then we will try to read `.items` out of an undefined object 😅

Comment (AarishMansur) [reply packages/zowe-explorer/src/trees/uss/UssFSProvider.ts:314]: oh yes I have missed that  just  added `?.` before `.items` and pushed the fix

**Review by traeok — COMMENTED** [review]
Hey @AarishMansur, thanks for working on this. The changes look good, but I noticed a changelog entry is missing. Can you please add one? I plan to approve the PR once that's addressed.

**Comment by AarishMansur** [comment]
@traeok  I have added changelog for this PR thanks for review

File: packages/zowe-explorer/CHANGELOG.md [hunk 2]
```diff
@@ -17,6 +17,7 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 - Added error handling for a "Could not list members" error that appears when deleting an expanded PDS using an SSH profile. [#768](https://github.com/zowe/zowex/issues/768)
 - Fixed an issue where valuable details in error messages were truncated, particularly when using SSH profiles. [zowex/#766](https://github.com/zowe/zowex/issues/766)
 - Fixed "Open Selected Data Set" action opening data sets in the z/OSMF tree instead of the originating extender profile's tree when invoked from a spool file. [#4372](https://github.com/zowe/zowe-explorer-vscode/issues/4372)
+- Fixed an issue where the USS directory-detection heuristic could crash when extenders return `items: undefined/null` or omit `apiResponse`, and added empty-list fallback paths for extender compatibility. [#4192](https://github.com/zowe/zowe-explorer-vscode/issues/4192)
```
Comment (anaxceron) [inline packages/zowe-explorer/CHANGELOG.md]: We should describe what `items: undefined/null` and `apiResponse` are: "... when extenders return the command/class/attribute/cat `items: undefined/null` or omit the flag/attribute/cat `apiResponse` ... "

Comment (AarishMansur) [reply packages/zowe-explorer/CHANGELOG.md]: Hey, just to clarify  
```ts
- Fixed an issue where the USS directory detection heuristic could crash when extenders return the `items` array property as `undefined`/`null` or omit the `apiResponse` response object, and added empty-list fallback paths for extender compatibility. [#4192](https://github.com/zowe/zowe-explorer-vscode/issues/4192)
```
 Does that work, or would you prefer a different wording in the changelog?

Comment (anaxceron) [reply packages/zowe-explorer/CHANGELOG.md]: @AarishMansur this is great, thank you 😸

Comment (AarishMansur) [reply packages/zowe-explorer/CHANGELOG.md]: @anaxceron  done i have pushed my changes thanks for review

**Review by jace-roell — CHANGES_REQUESTED** [review]
Thank you for the changes @AarishMansur! Left a quick nit-pick suggestion.

File: packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts [hunk 3]
```diff
@@ -171,10 +171,13 @@ export class ZoweUSSNode extends ZoweTreeNode implements IZoweUSSTreeNode {
             return undefined;
         }
         const response = await this.getUssFiles(this.profile);
-        if (response.success === false || response.apiResponse?.items.length === 0 || response.apiResponse?.items.length > 1) {
+        if (response.success === false || response.apiResponse?.items?.length === 0 || response.apiResponse?.items?.length > 1) {
```
Comment (jace-roell) [inline packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: (nit) this could collapse to `response.success === false || response.apiResponse?.items?.length !== 1)`

**Review by anaxceron — APPROVED** [review]

File: packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts (hunk 3 above)
Comment (AarishMansur) [reply packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: done pushed my changes

Comment (jace-roell) [reply packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: In this case the `.length === 0` check could be omitted, sorry for the confusion 😅

Comment (AarishMansur) [reply packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: oh no worries so to avoid further confusion the check should be `response.success === false || response.apiResponse?.items?.length !== 1` only, right? No `=== 0` or anything else needed?

Comment (jace-roell) [reply packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: Nope, thats it!

Comment (AarishMansur) [reply packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: done

**Review by jace-roell — APPROVED** [review]

File: packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts [hunk 4]
```diff
@@ -171,10 +171,13 @@ export class ZoweUSSNode extends ZoweTreeNode implements IZoweUSSTreeNode {
             return undefined;
         }
         const response = await this.getUssFiles(this.profile);
-        if (response.success === false || response.apiResponse?.items.length === 0 || response.apiResponse?.items.length > 1) {
+        if (response.success === false ||  response.apiResponse?.items?.length !== 1) {
+            return undefined;
+        }
+        const item = response.apiResponse.items?.[0];
```
Comment (traeok) [inline packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: At this point in the code, we have confirmed that `response.apiResponse.items` is non-null and has one item:


```suggestion
        const item = response.apiResponse.items[0];
```

**Review by traeok — COMMENTED** [review]
Thanks for working on this @AarishMansur! Had a quick suggestion around declaring `const item = ...` but aside from that, changes make sense. I plan to approve the PR once that is addressed.

File: packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts (hunk 4 above)
Comment (AarishMansur) [reply packages/zowe-explorer/src/trees/uss/ZoweUSSNode.ts]: done !

**Review by zFernand0 — APPROVED** [review]
LGTM! 😋

**Comment by AarishMansur** [comment]
Hey, I see 3 CI failures on the latest run:

1. **Audit** : pre-existing upstream vulnerabilities (`linkify-it`, `dompurify`), not from my changes
2. **Lint** : `vscode:prepublish` auto generated localization files from my changelog entry. Should I commit those?
3. **Coverage** :  patch at 88.89%, should I work on adding test to cover it 

Let me know how you'd like me to proceed. Thanks!

**Comment by zFernand0** [comment]
Hey @AarishMansur,

> Audit : pre-existing upstream vulnerabilities (linkify-it, dompurify), not from my changes

We will address these in a separate PR 🙏 

> Lint : vscode:prepublish auto generated localization files from my changelog entry. Should I commit those?

Please commit those changes 🙏 

> Coverage : patch at 88.89%, should I work on adding test to cover it

Ideally yes, you should aim for 90% or more 🙏

**Comment by AarishMansur** [comment]
@zFernand0 Hey I have added test cases and also commit my auto generated localization files can you rerun the cls ?

**Review by zFernand0 — APPROVED** [review]
LGTM! 😋 
---
Please avoid force-pushing going forward 🙏

File: packages/zowe-explorer/__tests__/__unit__/trees/uss/USSUtils.unit.test.ts [hunk 5]
```diff
@@ -350,4 +350,27 @@ describe("USSUtils Unit Tests - filterHiddenFiles", () => {
...
+        const mockResponse = {
+            success: true,
+            commandResponse: "",
+            apiResponse: { items: undefined },
+        } as any;
+
+        const result = await USSUtils.filterHiddenFiles(mockResponse);
+
+        expect(result.apiResponse.items).toEqual([]);
+    });
```
Comment (zFernand0) [inline packages/zowe-explorer/__tests__/__unit__/trees/uss/USSUtils.unit.test.ts:364]: The one line of coverage we are missing is if the first and only item in the `items` array is null or undefined
image

Can you add one more test that has the following mockResponse?

```
const mockResponse = {
            success: true,
            commandResponse: "",
            apiResponse: { items: [null] },
        } as any;
```
AND/OR
```
const mockResponse = {
            success: true,
            commandResponse: "",
            apiResponse: { items: [undefined] },
        } as any;
```

Comment (zFernand0) [reply packages/zowe-explorer/__tests__/__unit__/trees/uss/USSUtils.unit.test.ts:364]: We already had a test for `[null]`. 🤯 

Thanks for adding one for `[undefined]` 🙏  (31c3626)

codecov still complains about the `item == null` condition not being covered 🤔

Comment (AarishMansur) [reply packages/zowe-explorer/__tests__/__unit__/trees/uss/USSUtils.unit.test.ts:364]: @zFernand0  yes I cant understand why I still get 88.88 coverage could this be cached?

File: packages/zowe-explorer/__tests__/__unit__/trees/uss/ZoweUSSNode.unit.test.ts [hunk 6]
```diff
@@ -2020,6 +2020,23 @@ describe("ZoweUSSNode Unit Tests - Function node.fetchAttributes", () => {
...
+        const node = new ZoweUSSNode({
+            label: "testFile",
+            collapsibleState: vscode.TreeItemCollapsibleState.None,
+            profile: globalMocks.profileOne,
+        });
+        vi.spyOn(UssFSProvider.instance, "listFiles").mockResolvedValueOnce({
+            success: true,
+            apiResponse: { items: [undefined] },
+            commandResponse: "",
+        });
```
Comment (zFernand0) [inline packages/zowe-explorer/__tests__/__unit__/trees/uss/ZoweUSSNode.unit.test.ts]: we are spying and mocking the resolved value of  `listFiles` but we should be looking at the `getUssFiles` function

Image
